docs/decisions/0011-enterprise-subsidy-enrollments-and-entitlements.rst

```diff
@@ -0,0 +1,143 @@
+- converted_at (NULL DateTime).
+- (cached property) course_entitlement_id (query look up of related CourseEntitlement)
+
+TBD:
+-- A built in method of entitlement conversion to enrollment
+```
+
+Reworked and added table inheritance to all subsidy based enrollment tables. As such all subsidy based fulfillment records will have access to these fields:
+
+```
+EnterpriseFulfillmentSource
+---------------------------
+- uuid, created, modified, history (boilerplate)
+- fulfillment_ty (NOT NULL, char selection: (`license`, `learner_credit`, `coupon_code`))
+- enterprise_course_entitlement (NOT NULL, FK to EnterpriseCourseEntitlement)
+- enterprise_course_enrollment (NOT NULL, FK to EnterpriseCourseEnrollment)
+- is_revoked (Default False, Bool)
+```
+
+Models inheriting `EnterpriseFulfillmentSource`:
+
+```
+LicensedEnterpriseCourseEnrollment (inherited from EnterpriseFulfillmentSource)
+-------------------------------------------------------------------------------
+- license_uuid (NOT NULL, UUID field)
+```
+
+```
+LearnerCreditEnterpriseCourseEnrollment (inherited from EnterpriseFulfillmentSource)
+------------------------------------------------------------------------------------
+- transaction_id (NOT NULL, UUID field)
+```
+
+[NOTE] Even though these models are labeled as `...Enrollment`s, they can reference entitlements as well as enrollments. In fact, despite both `enterprise_course_entitlement` `enterprise_course_enrollment` both being nullable, there is validation on the `EnterpriseFulfillmentSource` which will guarantees one of these values must exist.
+
+To support interactions with these reworked and new models, we've buffed out the bulk enrollment (`enroll_learners_in_courses`) EnterpriseCustomerViewSet view to support subsidy enrollments. `enrollment_info` parameters supplied to the endpoint can now include transaction ID's that will detected and realized into a `LearnerCreditEnterpriseCourseEnrollment` record.
+
+**How we'd use this in code**
+
+```
+# In the parent class...
+@classmethod
+def get_fulfillment_source(cls, enrollment_id, entitlement_id=None):
+   return cls.objects.select_related(
+     # all child tables joined here
+   ).filter(
+     cls.enterprise_course_enrollment=enrollment_id
+   )
+   # do kwargs stuff here to optionally pass in a non-null
+   # entitlement id to filter by...
+
+@property
+def fulfillment_status(self):
+  if not self.enterprise_course_enrollment:
+    return 'entitled'
+  return 'enrolled'
+```
+
+# Consequences
+
+- Table inheritance means that we’ll most likely have to do JOINs in our code and in our analytics/reporting.
+
+- There exists a subsidy based enrollment table already (`LicensedEnterpriseCourseEnrollment`), this table and it's records will need to be migrated to the table inheritance structure which would complicate our django migration hierarchy.
+
+# Further Improvements
+
+- Verify transaction ID's are real on creation through the bulk enrollment endpoint
```

Comment:
tl;dr how we're mitigating this for now: for learner credit, requests to fulfill transactions will only come from an authenticated _system_ user (i.e. the `enterprise-subsidy_worker` user or whatever).